### DISCUSSION TITLE
Self hosting instructions

### DISCUSSION THREAD

**Discussion by danboid** [antonym/netboot.xyz#1092] [Q&A]
I would like to set up a self-hosted netboot.xyz server so that I can try creating my own custom entries. netboot.xyz used to have manual instructions for doing this but now they've been replaced by an ansible playbook (that doesn't work for me) and a docker image it seems.

So I end up on this page:

https://netboot.xyz/docs/selfhosting

IMO this guide gets to the "Deploying using Ansible" instructions too soon as it hasn't answered several important questions:

* What OSes can netboot.xyz be deployed too? Any Linux/BSD? Any OS that can run ansible and python? It seems it must be a 64 bit Linux variant?

* What is the minimum recommended spec of a netboot.xyz server? Is an ethernet connection required or can it work via wifi? Does it need to be a gigabit connection? What ports need to be open?

* Does the user need to configure a static IP address on the target machine before they run the playbook?

* Does it matter if there is already a (proxy) DHCP server on the network? How will running the netboot.xyz playbook or installing the container affect the users network when it is running?

I tried running the playbook on my rather ancient bananapi running armbian bullseye 22.02 and the end of the install looked like this:

```
TASK [netbootxyz : Install required packages] ****************************************************************************************************
ok: [localhost] => (item=apache2)
changed: [localhost] => (item=binutils-dev)
failed: [localhost] (item=binutils-aarch64-linux-gnu) => {"ansible_loop_var": "item", "changed": false, "item": "binutils-aarch64-linux-gnu", "msg": "No package matching 'binutils-aarch64-linux-gnu' is available"}
ok: [localhost] => (item=dosfstools)
ok: [localhost] => (item=figlet)
failed: [localhost] (item=gcc-aarch64-linux-gnu) => {"ansible_loop_var": "item", "changed": false, "item": "gcc-aarch64-linux-gnu", "msg": "No package matching 'gcc-aarch64-linux-gnu' is available"}
ok: [localhost] => (item=build-essential)
changed: [localhost] => (item=genisoimage)
ok: [localhost] => (item=git)
changed: [localhost] => (item=isolinux)
changed: [localhost] => (item=liblzma-dev)
changed: [localhost] => (item=libslirp-dev)
failed: [localhost] (item=syslinux) => {"ansible_loop_var": "item", "changed": false, "item": "syslinux", "msg": "No package matching 'syslinux' is available"}
changed: [localhost] => (item=syslinux-common)
ok: [localhost] => (item=toilet)

PLAY RECAP ***************************************************************************************************************************************
localhost                  : ok=16   changed=7    unreachable=0    failed=1    skipped=10   rescued=0    ignored=0   
```

It looks to me like netboot.xyz doesn't support creating the disk images under 32 bit Linux? I didn't want disk images anyway, I wanted the playbook to set up a local netboot.xyz server but it doesn't seem to have done that or at least I was unable to PXE boot my laptop after the playbook had finished. 

Have I got to use the docker image to set up a local netboot.xyz server or have I got to run additional commands / configure my DHCP server manually? Do the playbook and docker image serve the same purpose but in different ways?

I would like to see all of these questions answered on the self hosting page because at the moment its quite vague what the playbook and docker container actually do and how we are supposed to use them.

**Comment by danboid** [comment]
I presume that to get what I want I would also need to follow the steps on:

https://netboot.xyz/docs/booting/tftp/

If so, it should say so on the self hosting page. I expected that the playbook was going to do all of the steps listed on the tftp boot page but maybe thats only for making boot disks?

**Reply by danboid** [reply]
What happened to the instructions for manually setting up netboot.xyz? I've had a look on the wayback machine but I can't find them.

I want to install netboot.xyz on an ARM based TV box. No docker.

Thanks

**Comment by danboid** [comment]
There is am ARM docker container for netboot.xyz here:

https://hub.docker.com/r/linuxserver/netbootxyz

Like the official container, it also doesn't include a DHCP server so saying that netboot.xyz can be set up with one command is a lie, which is what is claimed at the start of the docker page for netboot.

Why does the docker image not use dnsmasq for networking? This would make adding DHCP easier, dnsmasq could handle the TFTP and DHCP/DNS, then maybe the one command self hosted set up would be possible.

**Reply by samip5** [reply]
That container is to be deprecated in favor of the netbootxyz/docker-netbootxyz.

Let me guess, you are running arm 32bit? That's awfully unsupported platform. You are recommended to run arm64.
https://github.com/netbootxyz/docker-netbootxyz/blob/master/.github/workflows/build.yml#L114

**Reply by danboid** [reply]
No, its Armbian Ubuntu arm64 and I've got it working now. The example DHCP config on the docker page doesn't work.

**Reply by danboid** [reply]
https://github.com/netbootxyz/netboot.xyz/issues/1093

**Comment by danboid** [comment]
I have tried using both the official netboot.xyz container (which I'm surprised installed on my ARM device) and the linuxserver netboot.xyz (which claims to support ARM64) but it both cases I get the same error:

```
PXE-T01: File not found
PXE-E3B: TFTP Error - File not found
```

This is the command I used to start the linuxserver netbootxyz:

```
docker run -d \
  --name=netbootxyz \
  -p 3000:3000                       `# sets webapp port` \
  -p 69:69/udp                       `# sets tftp port` \
  -p 8080:80                         `# optional` \
  --restart unless-stopped \
  linuxserver/netbootxyz
```

I used a similar command to start the official netboot container.

Here is my DHCP server config:

```
root@armbian:/etc/dhcp# cat dhcpd.conf
option arch code 93 = unsigned integer 16;

subnet 192.168.0.0 netmask 255.255.255.0 {
  range 192.168.0.34 192.168.0.254;
  next-server 192.168.0.33;
  option subnet-mask 255.255.255.0;
  option routers 192.168.0.1;
  option broadcast-address 192.168.0.255;
  option domain-name "mynetwork.lan";
  option domain-name-servers 1.1.1.1;
  if exists user-class and ( option user-class = "iPXE" ) {
    filename "http://boot.netboot.xyz/menu.ipxe";
  } elsif option arch = encode-int ( 16, 16 ) {
    filename "http://boot.netboot.xyz/ipxe/netboot.xyz.efi";
    option vendor-class-identifier "HTTPClient";
  } elsif option arch = 00:07 {
    filename "netboot.xyz.efi";
  } elsif option arch = 00:00 {
    filename "netboot.xyz.pxe";
  } else {
    filename "netboot.xyz.kpxe";
  }
}
```

I think the example DHCP config should be more like this to match most peoples home networks, using the 192.168 private network.

The netboot docker page states:

"The main configuration you will need to change are next-server and filename/boot-file-name."

Why should the user need to configure the filename/boot-file-name? Isn't this already configured in the docker container? If so then the file names given in the example DHCP config should work unmodified and the only address thety should need to change is the next-server address to match their docker/DHCP server.

**Comment by danboid** [comment]
I have got my ARM netboot.xyz server running. The example dhcpd config given doesn't work but this works for me ( `/etc/dhcp/dhcpd.conf` ): 

```
option arch code 93 = unsigned integer 16;

subnet 192.168.0.0 netmask 255.255.255.0 {
  range 192.168.0.34 192.168.0.254;
  next-server 192.168.0.33;
  option subnet-mask 255.255.255.0;
  option routers 192.168.0.1;
  option broadcast-address 192.168.0.255;
  option domain-name "mynetwork.lan";
  option domain-name-servers 1.1.1.1;
  filename "netboot.xyz.kpxe";
}
```

**Comment by danboid** [comment]
To answer one of my own questions, yes the official netboot.xyz docker container does work under ARM Linux.
